aten/src/ATen/native/cuda/RNN.cu

```diff
@@ -0,0 +1,601 @@
+  auto hidden_gatesI = getTensorInfo<scalar_t, index_type>(hidden_gates);
+  auto input_biasI = tryGetTensorInfo<scalar_t, index_type>(input_bias);
+  auto hidden_biasI = tryGetTensorInfo<scalar_t, index_type>(hidden_bias);
+  auto hxI = getTensorInfo<scalar_t, index_type>(hx);
+  auto hyI = getTensorInfo<scalar_t, index_type>(hy);
+  auto workspaceI = getTensorInfo<scalar_t, index_type>(workspace);
+  index_type hidden_size = hxI.sizes[hxI.dims-1];
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  if (allContiguous({input_gates, hidden_gates, input_bias, hidden_bias, hx, hy, workspace})) {
+    collapseDims(input_gatesI, hidden_gatesI, input_biasI, hidden_biasI, hxI, hyI, workspaceI);
+    kernel::gru_cell_forward<scalar_t, accscalar_t, index_type, 1>
+      <<<grid, block, 0, stream>>>
+        (input_gatesI, hidden_gatesI, input_biasI, hidden_biasI, hxI, hyI, workspaceI, hidden_size, numel);
+  } else {
+    kernel::gru_cell_forward<scalar_t, accscalar_t, index_type, 2>
+      <<<grid, block, 0, stream>>>
+        (input_gatesI, hidden_gatesI, input_biasI, hidden_biasI, hxI, hyI, workspaceI, hidden_size, numel);
+  }
+}
+
+template<typename scalar_t, typename index_type>
+void gru_backward_index(const Tensor& grad_hy, const Tensor& workspace,
+                        const Tensor& grad_input_gates, const Tensor& grad_hidden_gates, const Tensor& grad_hx) {
+  using accscalar_t = acc_type<scalar_t, /*is_cuda=*/true>;
+
+  dim3 block, grid;
+  int64_t numel = grad_hy.numel();
+  getLaunchConfig(block, grid, numel);
+
+  auto grad_hyI = getTensorInfo<scalar_t, index_type>(grad_hy);
+  auto workspaceI = getTensorInfo<scalar_t, index_type>(workspace);
+  auto grad_input_gatesI = getTensorInfo<scalar_t, index_type>(grad_input_gates);
+  auto grad_hidden_gatesI = getTensorInfo<scalar_t, index_type>(grad_hidden_gates);
+  auto grad_hxI = getTensorInfo<scalar_t, index_type>(grad_hx);
+  index_type hidden_size = grad_hyI.sizes[grad_hyI.dims-1];
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  if (allContiguous({grad_hy, workspace, grad_input_gates, grad_hidden_gates, grad_hx})) {
+    collapseDims(grad_hyI, workspaceI, grad_input_gatesI, grad_hidden_gatesI, grad_hxI);
+    kernel::gru_cell_backward<scalar_t, accscalar_t, index_type, 1>
+      <<<grid, block, 0, stream>>>
+        (grad_input_gatesI, grad_hidden_gatesI, grad_hyI, grad_hxI, workspaceI, hidden_size, numel);
+  } else {
+    kernel::gru_cell_backward<scalar_t, accscalar_t, index_type, 2>
+      <<<grid, block, 0, stream>>>
+        (grad_input_gatesI, grad_hidden_gatesI, grad_hyI, grad_hxI, workspaceI, hidden_size, numel);
+  }
+}
+
+} // anonymous namespace
+
+// Note [64-bit index math check elision]
+// It's enough to perform the check for 64-bit math on the largest tensor only.
+// If 32-bit is enough for it, it will suffice for all other tensors too, and we
+// can save some work using this trick.
```

Comment:
Remind me why `input_gates` is guaranteed to be bigger than `cx`?

Reply:
Because `input_gates` is effectively `cx` repeated `num_gates` times (3 for GPU, 4 for LSTM).